lnrpc/routerrpc/router_server.go

```diff
@@ -398,14 +441,350 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	// the last hop's fee amount has to be sent across the payment path we
+	// have to add it to the original payment amount. Only then will the
+	// payment probe be able to determine the correct fee to the last hop
+	// prior to the private destination. For example, if the user wants to
+	// send 1000 sats to a private destination and the last hop's fee is 10
+	// sats, then 1010 sats will have to arrive at the last hop. This means
+	// that the probe has to be dispatched with 1010 sats to correctly
+	// calculate the routing fee.
+	//
+	// Calculate the hop fee for the last hop manually.
+	hopFee := lspHopHint.HopFee(*payReq.MilliSat)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add the last hop's fee to the requested payment amount that we want
+	// to get an estimate for.
+	probeRequest.AmtMsat += int64(hopFee)
+
+	// Dispatch the payment probe with adjusted fee amount.
+	resp, err := s.sendProbePayment(ctx, probeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// The probe succeeded, so we can add the last hop's fee to fee the
+	// payment probe returned.
+	resp.RoutingFeeMsat += int64(hopFee)
+
+	lastHopTimeLock := int64(lspHopHint.CLTVExpiryDelta)
+	if lastHopTimeLock < int64(payReq.MinFinalCLTVExpiry()) {
+		lastHopTimeLock = int64(payReq.MinFinalCLTVExpiry())
+	}
+	resp.TimeLockDelay += lastHopTimeLock
+
+	return resp, nil
+}
+
+// isLSP checks if the route hints indicate an LSP. An LSP is indicated in the
+// first return value if the last node in each route hint has the same node id.
+// If an LSP is indicated, the remaining hops in each route hint are returned
+// to the caller to be considered by the caller as route hints to the LSP node.
+// Additionally, the identified LSP node is returned as a separate hop hint.
+// This hop hint also contains the maximum fee setting if LSP hop hints contain
+// different fee settings. This is to heuristically determine a worst-case
+// estimate.
+func isLSP(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) (bool,
+	[][]zpay32.HopHint, *zpay32.HopHint) {
+
+	if len(routeHints) == 0 || len(routeHints[0]) == 0 {
+		return false, [][]zpay32.HopHint{}, nil
+	}
+
+	// In case we are dealing with an LSP, we construct a modified list of
+	// route hints that allows the caller to probe the LSP, which itself is
+	// returned as a separate hop hint.
+	adjustedRouteHints := make([][]zpay32.HopHint, 0, len(routeHints))
+
+	// If the final pubkey of each route hint is the same, we can assume
+	// that this identifies an LSP node. The first candidate is our
+	// reference hop hint.
+	refHopHint := routeHints[0][len(routeHints[0])-1]
+
+	// Strip off the potential LSP hop hint from the reference route hint,
+	// and add the remaining hop hints to the adjusted route hints.
+	if len(routeHints[0]) > 1 {
+		adjustedRouteHints = append(
+			adjustedRouteHints,
+			routeHints[0][:len(routeHints[0])-1],
+		)
+	}
+	for i := 1; i < len(routeHints); i++ {
```

Comment:
I'd like to see some of this broken up, into the separate checks. Right now all of the different heuristics are mixed together in the same loop which makes it hard to figure out what is being looked at. Instead, I'd recommend having a "node id homogeneity check", a "max fee selector", a "max cltv selector", etc. and then have the overall isLSP function be a list of these heuristics bundled. Especially with early exit, reasoning about for loops larger than a screen height is difficult, and when someone returns to these heuristics months from now they'll be grateful that things are neatly organized into their constituent pieces.

Reply:
I don't think there's a ton of value to doing a golang interface for `heuristicCheck` but what I was saying is that we have three distinct things I was able to see. I'll provide signatures

```go
// note it is possible not all of these individual heuristics require all of these arguments, and
// the arguments should be trimmed if unused.

func HasHomogenousNodeIDs(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) bool { ... }

func FindMaxCLTV(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) uint32 { ... }

func FindMaxFee(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) lnwire.MilliSatoshi { ... }

func isLSP(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) *lspHeuristics {
    var heuristics lspHeuristics
    heuristics.sameNodeID = HasHomogenousNodeIDs(routeHints, amt)
    heuristics.maxCLTV = FindMaxCLTV(routeHints, amt)
    heuristics.maxFee = FindMaxFee(routeHints, amt)

    // maybe we do some further computation on these results together here in the future or something
}
```

This way we can look at `isLSP` and be like "Oh 🤔, that's what we are measuring to make our determinations", and each of those heuristics *only* measures one thing, making it easy to follow and verify it works correctly as designed. This way, if we want to add a new check, it's simply a matter of implementing a new standalone function for it, and then adding it to the bundle for `isLSP`, rather than trying to splice it into a large for-loop.

Hopefully this sheds light on the approach I'm going for? Lmk if you foresee issues here, but I think it will make the code substantially nicer to look at 😁

Reply:
Thanks that makes total sense. I am glad I asked as I was taking your "isLsp is a list of these heuristics" too literal by passing in a slice of these checks.

I'll go with your suggestion though I think the checks don't need to be exported.